(Android - Crash): FragmentManager is already executing transactions

<!--
NOTE: please submit only bug reports here, any new questions or feature requests should be submitted in Discussions:
https://github.com/software-mansion/react-native-screens/discussions
 -->

## Description
When react native screens are enabled on Android and the architecture is based on a single `Activity` plus React native is loaded in a Fragment, there's a crash when navigating and restoring the fragment state.
<!--
Tell us what's happening here.
-->
Having a single `Activity` app and using `Fragments` with navigation components as describe [here](https://developer.android.com/guide/navigation/navigation-getting-started) brings us to load React Native apps on a `Fragment`. We also want to enable React native screens, however, when enabled on Android, there's the following crash:
`    java.lang.RuntimeException: Unable to resume activity {com.swmansion.rnscreens.example/com.swmansion.rnscreens.example.MainActivity}: java.lang.IllegalStateException: FragmentManager is already executing transactions
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4445)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4477)
        at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2067)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7697)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:952)
     Caused by: java.lang.IllegalStateException: FragmentManager is already executing transactions
        at androidx.fragment.app.FragmentManager.ensureExecReady(FragmentManager.java:1686)
        at androidx.fragment.app.FragmentManager.execSingleAction(FragmentManager.java:1716)
        at androidx.fragment.app.BackStackRecord.commitNowAllowingStateLoss(BackStackRecord.java:323)
        at com.swmansion.rnscreens.ScreenContainer.removeMyFragments(ScreenContainer.kt:216)
        at com.swmansion.rnscreens.ScreenContainer.onDetachedFromWindow(ScreenContainer.kt:228)
        at android.view.View.dispatchDetachedFromWindow(View.java:20538)
        at android.view.ViewGroup.dispatchDetachedFromWindow(ViewGroup.java:3942)
        at android.view.ViewGroup.dispatchDetachedFromWindow(ViewGroup.java:3934)
        at android.view.ViewGroup.dispatchDetachedFromWindow(ViewGroup.java:3934)
        at android.view.ViewGroup.removeViewInternal(ViewGroup.java:5560)
        at android.view.ViewGroup.removeViewInternal(ViewGroup.java:5531)
        at android.view.ViewGroup.removeView(ViewGroup.java:5462)
        at androidx.fragment.app.FragmentContainerView.removeView(FragmentContainerView.kt:288)
        at androidx.fragment.app.SpecialEffectsController$Operation$State.applyState(SpecialEffectsController.java:470)
        at androidx.fragment.app.DefaultSpecialEffectsController.applyContainerChanges(DefaultSpecialEffectsController.java:822)
        at androidx.fragment.app.DefaultSpecialEffectsController$1.run(DefaultSpecialEffectsController.java:111)
        at androidx.fragment.app.SpecialEffectsController$Operation.complete(SpecialEffectsController.java:713)
        at androidx.fragment.app.SpecialEffectsController$FragmentStateManagerOperation.complete(SpecialEffectsController.java:770)
        at androidx.fragment.app.SpecialEffectsController$Operation.completeSpecialEffect(SpecialEffectsController.java:690)
        at androidx.fragment.app.DefaultSpecialEffectsController$SpecialEffectsInfo.completeSpecialEffect(DefaultSpecialEffectsController.java:855)
        at androidx.fragment.app.DefaultSpecialEffectsController.startAnimations(DefaultSpecialEffectsController.java:154)
        at androidx.fragment.app.DefaultSpecialEffectsController.executeOperations(DefaultSpecialEffectsController.java:123)
        at androidx.fragment.app.SpecialEffectsController.executePendingOperations(SpecialEffectsController.java:306)
        at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:1903)
        at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:1814)
        at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:1751)
        at androidx.fragment.app.Fragment.performResume(Fragment.java:3177)
        at androidx.fragment.app.FragmentStateManager.resume(FragmentStateManager.java:606)
        at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:285)
2022-07-03 22:00:08.008 30152-30152/com.swmansion.rnscreens.example E/AndroidRuntime:     at androidx.fragment.app.FragmentStore.moveToExpectedState(FragmentStore.java:113)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1424)
        at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:2968)
        at androidx.fragment.app.FragmentManager.dispatchResume(FragmentManager.java:2900)
        at androidx.fragment.app.FragmentController.dispatchResume(FragmentController.java:285)
        at androidx.fragment.app.FragmentActivity.onResumeFragments(FragmentActivity.java:334)
        at androidx.fragment.app.FragmentActivity.onPostResume(FragmentActivity.java:323)
        at androidx.appcompat.app.AppCompatActivity.onPostResume(AppCompatActivity.java:204)
        at android.app.Activity.performResume(Activity.java:8160)
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4435)`

The reason of the crash seems to be:

- When setting up `ScreenContainer` it receives the `FragmentManager` from the Activity at some point.
- This same `FragmentManager` is used by the Navigation Components to navigate between fragments.
- When the `Fragment` that has react native tries to recover the state after a navigation, it tries to use the `Activity` `FragmentManager` which is already in use.

### Screenshots

## Steps To Reproduce
To reproduce, I've recreated in the `Example` app a common use case: Load React native, put the app in the background and trigger a deep link. I've prepared the `Example` to do this so the steps are:

1. Launch the Example app
2. Tap Stack Presentation
3. Tap Form Sheet
4. Put some values in the form.
5. Background the app
6. Launch the following adb comand: `adb shell am start -W -a android.intent.action.VIEW -d "screens://screens" com.swmansion.rnscreens.example`

### Expected behavior
The react native screen should display without a crash

### Actual behavior
Crash with the above stack trace
## Reproduction
To reproduce the crash I've forked  `react-native-screens` github project and modified the example with the described architecture. The project can be found here: https://github.com/luislukas/react-native-screens

I reported this issue originally on https://github.com/software-mansion/react-native-screens/issues/819 but due to our architecture we had to disable `react-native-screens`. We would like to enable them now so any hint/help would be appreciated.

<!--
Please provide a Snack ([https://snack.expo.io/](https://snack.expo.io/)), link to a GitHub repository or provide a minimal code example that reproduces the problem. Issues without valid reproduction will be closed after a specific period.
Here are some tips for providing a minimal example: [https://stackoverflow.com/help/mcve](https://stackoverflow.com/help/mcve).
-->

## Platform
<!--
Please list platforms that the issue is affecting.
-->

- [ ] iOS
- [X] Android
- [ ] Web
- [ ] Windows
- [ ] tvOS

## Architecture
I've forked the `react-native-screens` and added the Navigation Componentes plus `Fragments` architecture described aboved.
<!-- 
Please check the appropriate boxes with architectures the issue is specific to. 
 -->

- [ ] Paper
- [ ] Fabric

## Workflow

<!-- 
Are you using Expo to stream you development? Please check "Managed workflow" checkbox below.
If your project uses bare react-native, please check "Bare workflow" checkbox.

Don't know? Check these docs: https://docs.expo.dev/introduction/managed-vs-bare/
-->

- [ ] Managed workflow
- [ ] Bare workflow

## Package versions
I'm using a fork of this project just changing the Android project architecture.

<!--
What are the exact versions of packages that you are using?

Remove packages from the table that you're not using or list additional if relevant.
-->

| package                                | version |
| -------------------------------------- | ------- |
| react-native                           |         |
| @react-navigation/native               |         |
| @react-navigation/native-stack         |         |
| react-native-screens                   |         |
| react-native-safe-area-context         |         |
| react-native-gesture-handler           |         |
| react-native-reanimated                |         |
| expo                                   |         |

--- a/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
+++ b/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
@@ -177,8 +177,17 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         while (context !is FragmentActivity && context is ContextWrapper) {
             context = context.baseContext
         }
+
         check(context is FragmentActivity) { "In order to use RNScreens components your app's activity need to extend ReactActivity" }
-        setFragmentManager(context.supportFragmentManager)
+
+        // In case React Native is loaded on a Fragment (not directly in activity) we need to find fragment manager
+        // whose root view is ReactRootView. As of now, we detect such case by checking whether any fragments are attached
+        // to activity which hosts ReactRootView.
+        if (context.supportFragmentManager.fragments.isNotEmpty()) {
+            setFragmentManager(resolveChildFragmentManagerOfReactRootView(context.supportFragmentManager)!!)
+        } else {
+            setFragmentManager(context.supportFragmentManager)
+        }
     }
 
     protected fun createTransaction(): FragmentTransaction {

--- a/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
+++ b/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
@@ -178,7 +178,7 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         // Otherwise we expect to connect directly with root view and get root fragment manager
         if (parent is Screen) {
             val screenFragment = parent.fragment
-            check(screenFragment != null) { "Parent Screen does not have its Fragment attached" }
+            checkNotNull(screenFragment) { "Parent Screen does not have its Fragment attached" }
             mParentScreenFragment = screenFragment
             screenFragment.registerChildScreenContainer(this)
             setFragmentManager(screenFragment.childFragmentManager)

--- a/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
+++ b/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
@@ -136,16 +136,24 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
     }
 
     private fun findReactRootViewChildVM(fragmentManager: FragmentManager): FragmentManager {
-        for (fragment in fragmentManager.fragments) {
-            return if (fragment.view is ReactRootView) {
-                fragment.childFragmentManager
-            } else {
-                findReactRootViewChildVM(fragment.childFragmentManager)
+        fun _findReactRootViewChildVM(fragmentManager: FragmentManager): FragmentManager? {
+            for (fragment in fragmentManager.fragments) {
+                return if (fragment.view is ReactRootView) {
+                    fragment.childFragmentManager
+                } else {
+                    findReactRootViewChildVM(fragment.childFragmentManager)
+                }
             }
+            return null
         }
 
-        // In case there are no hosted fragments we just return passed fragment manager
-        return fragmentManager;
+        return if (fragmentManager.fragments.isEmpty()) {
+            fragmentManager
+        } else {
+            checkNotNull(_findReactRootViewChildVM(fragmentManager)) {
+                "Failed to resolve fragment manager for ScreenContainer"
+            }
+        }
     }
 
     private fun setupFragmentManager() {

--- a/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
+++ b/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
@@ -135,13 +135,13 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         performUpdatesNow()
     }
 
-    private fun findReactRootViewChildVM(fragmentManager: FragmentManager): FragmentManager {
-        fun _findReactRootViewChildVM(fragmentManager: FragmentManager): FragmentManager? {
+    private fun findReactRootViewChildFM(fragmentManager: FragmentManager): FragmentManager {
+        fun _findReactRootViewChildFM(fragmentManager: FragmentManager): FragmentManager? {
             for (fragment in fragmentManager.fragments) {
                 return if (fragment.view is ReactRootView) {
                     fragment.childFragmentManager
                 } else {
-                    _findReactRootViewChildVM(fragment.childFragmentManager)
+                    _findReactRootViewChildFM(fragment.childFragmentManager)
                 }
             }
             return null
@@ -150,7 +150,7 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         return if (fragmentManager.fragments.isEmpty()) {
             fragmentManager
         } else {
-            checkNotNull(_findReactRootViewChildVM(fragmentManager)) {
+            checkNotNull(_findReactRootViewChildFM(fragmentManager)) {
                 "Failed to resolve fragment manager for ScreenContainer"
             }
         }
@@ -193,7 +193,7 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         // In case React Native is loaded on a Fragment (not directly in activity) we need to find fragment manager
         // whose fragment's view is ReactRootView. As of now, we detect such case by checking whether any fragments are attached
         // to activity which hosts ReactRootView.
-        setFragmentManager(findReactRootViewChildVM(context.supportFragmentManager))
+        setFragmentManager(findReactRootViewChildFM(context.supportFragmentManager))
     }
 
     protected fun createTransaction(): FragmentTransaction {

--- a/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
+++ b/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
@@ -135,6 +135,17 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         performUpdatesNow()
     }
 
+    private fun resolveChildFragmentManagerOfReactRootView(fragmentManager: FragmentManager): FragmentManager? {
+        for (fragment in fragmentManager.fragments) {
+            return if (fragment.view is ReactRootView) {
+                fragment.childFragmentManager
+            } else {
+                resolveChildFragmentManagerOfReactRootView(fragment.childFragmentManager)
+            }
+        }
+        return null;
+    }
+
     private fun setupFragmentManager() {
         var parent: ViewParent = this
         // We traverse view hierarchy up until we find screen parent or a root view

--- a/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
+++ b/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
@@ -135,24 +135,32 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         performUpdatesNow()
     }
 
-    private fun findReactRootViewChildFM(fragmentManager: FragmentManager): FragmentManager {
-        fun _findReactRootViewChildFM(fragmentManager: FragmentManager): FragmentManager? {
-            for (fragment in fragmentManager.fragments) {
-                return if (fragment.view is ReactRootView) {
-                    fragment.childFragmentManager
-                } else {
-                    _findReactRootViewChildFM(fragment.childFragmentManager)
-                }
-            }
-            return null
+    private fun findFragmentManagerForReactRootView(rootView: ReactRootView): FragmentManager {
+        var context = rootView.context
+
+        // ReactRootView is expected to be initialized with the main React Activity as a context but
+        // in case of Expo the activity is wrapped in ContextWrapper and we need to unwrap it
+        while (context !is FragmentActivity && context is ContextWrapper) {
+            context = context.baseContext
         }
 
-        return if (fragmentManager.fragments.isEmpty()) {
-            fragmentManager
+        check(context is FragmentActivity) {
+            "In order to use RNScreens components your app's activity need to extend ReactActivity"
+        }
+
+        // In case React Native is loaded on a Fragment (not directly in activity) we need to find
+        // fragment manager whose fragment's view is ReactRootView. As of now, we detect such case by
+        // checking whether any fragments are attached to activity which hosts ReactRootView.
+        return if (context.supportFragmentManager.fragments.isEmpty()) {
+            // We are in standard React Native application w/o custom native navigation based on fragments.
+           context.supportFragmentManager
         } else {
-            checkNotNull(_findReactRootViewChildFM(fragmentManager)) {
-                "Failed to resolve fragment manager for ScreenContainer"
-            }
+            // We are in some custom setup & we want to use the closest fragment manager in hierarchy.
+            // `findFragment` method throws IllegalStateException when it fails to resolve appropriate
+            // fragment. It might happen when e.g. React Native is loaded directly in Activity
+            // but some custom fragments are still used. However such use case seems highly unlikely
+            // so, as for now, we let application crash.
+            FragmentManager.findFragment<Fragment>(rootView).childFragmentManager
         }
     }
 
@@ -172,28 +180,15 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
             mParentScreenFragment = screenFragment
             screenFragment.registerChildScreenContainer(this)
             setFragmentManager(screenFragment.childFragmentManager)
-            return
+        } else {
+            // we expect top level view to be of type ReactRootView, this isn't really necessary but in
+            // order to find root view we test if parent is null. This could potentially happen also when
+            // the view is detached from the hierarchy and that test would not correctly indicate the root
+            // view. So in order to make sure we indeed reached the root we test if it is of a correct type.
+            // This allows us to provide a more descriptive error message for the aforementioned case.
+            check(parent is ReactRootView) { "ScreenContainer is not attached under ReactRootView" }
+            setFragmentManager(findFragmentManagerForReactRootView(parent))
         }
-
-        // we expect top level view to be of type ReactRootView, this isn't really necessary but in
-        // order to find root view we test if parent is null. This could potentially happen also when
-        // the view is detached from the hierarchy and that test would not correctly indicate the root
-        // view. So in order to make sure we indeed reached the root we test if it is of a correct type.
-        // This allows us to provide a more descriptive error message for the aforementioned case.
-        check(parent is ReactRootView) { "ScreenContainer is not attached under ReactRootView" }
-        // ReactRootView is expected to be initialized with the main React Activity as a context but
-        // in case of Expo the activity is wrapped in ContextWrapper and we need to unwrap it
-        var context = parent.context
-        while (context !is FragmentActivity && context is ContextWrapper) {
-            context = context.baseContext
-        }
-
-        check(context is FragmentActivity) { "In order to use RNScreens components your app's activity need to extend ReactActivity" }
-
-        // In case React Native is loaded on a Fragment (not directly in activity) we need to find fragment manager
-        // whose fragment's view is ReactRootView. As of now, we detect such case by checking whether any fragments are attached
-        // to activity which hosts ReactRootView.
-        setFragmentManager(findReactRootViewChildFM(context.supportFragmentManager))
     }
 
     protected fun createTransaction(): FragmentTransaction {

--- a/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
+++ b/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
@@ -155,14 +155,18 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         // must be treated separately.
         return if (context.supportFragmentManager.fragments.isEmpty()) {
             // We are in standard React Native application w/o custom native navigation based on fragments.
-           context.supportFragmentManager
+            context.supportFragmentManager
         } else {
             // We are in some custom setup & we want to use the closest fragment manager in hierarchy.
             // `findFragment` method throws IllegalStateException when it fails to resolve appropriate
             // fragment. It might happen when e.g. React Native is loaded directly in Activity
-            // but some custom fragments are still used. However such use case seems highly unlikely
-            // so, as for now, we let application crash.
-            FragmentManager.findFragment<Fragment>(rootView).childFragmentManager
+            // but some custom fragments are still used. Such use case seems highly unlikely
+            // so, as for now we fallback to activity's FragmentManager in hope for the best.
+            try {
+                FragmentManager.findFragment<Fragment>(rootView).childFragmentManager
+            } catch (ex: IllegalStateException) {
+                context.supportFragmentManager
+            }
         }
     }
 
@@ -177,11 +181,13 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         // If parent is of type Screen it means we are inside a nested fragment structure.
         // Otherwise we expect to connect directly with root view and get root fragment manager
         if (parent is Screen) {
-            checkNotNull(parent.fragment?.let { screenFragment ->
-                mParentScreenFragment = screenFragment
-                screenFragment.registerChildScreenContainer(this)
-                setFragmentManager(screenFragment.childFragmentManager)
-            }) { "Parent Screen does not have its Fragment attached" }
+            checkNotNull(
+                parent.fragment?.let { screenFragment ->
+                    mParentScreenFragment = screenFragment
+                    screenFragment.registerChildScreenContainer(this)
+                    setFragmentManager(screenFragment.childFragmentManager)
+                }
+            ) { "Parent Screen does not have its Fragment attached" }
         } else {
             // we expect top level view to be of type ReactRootView, this isn't really necessary but in
             // order to find root view we test if parent is null. This could potentially happen also when

--- a/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
+++ b/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
@@ -135,12 +135,12 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         performUpdatesNow()
     }
 
-    private fun resolveChildFragmentManagerOfReactRootView(fragmentManager: FragmentManager): FragmentManager? {
+    private fun findReactRootViewChildVM(fragmentManager: FragmentManager): FragmentManager? {
         for (fragment in fragmentManager.fragments) {
             return if (fragment.view is ReactRootView) {
                 fragment.childFragmentManager
             } else {
-                resolveChildFragmentManagerOfReactRootView(fragment.childFragmentManager)
+                findReactRootViewChildVM(fragment.childFragmentManager)
             }
         }
         return null;
@@ -184,7 +184,7 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         // whose root view is ReactRootView. As of now, we detect such case by checking whether any fragments are attached
         // to activity which hosts ReactRootView.
         if (context.supportFragmentManager.fragments.isNotEmpty()) {
-            setFragmentManager(resolveChildFragmentManagerOfReactRootView(context.supportFragmentManager)!!)
+            setFragmentManager(findReactRootViewChildVM(context.supportFragmentManager)!!)
         } else {
             setFragmentManager(context.supportFragmentManager)
         }

--- a/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
+++ b/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
@@ -151,6 +151,8 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         // In case React Native is loaded on a Fragment (not directly in activity) we need to find
         // fragment manager whose fragment's view is ReactRootView. As of now, we detect such case by
         // checking whether any fragments are attached to activity which hosts ReactRootView.
+        // See: https://github.com/software-mansion/react-native-screens/issues/1506 on why the cases
+        // must be treated separately.
         return if (context.supportFragmentManager.fragments.isEmpty()) {
             // We are in standard React Native application w/o custom native navigation based on fragments.
            context.supportFragmentManager

--- a/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
+++ b/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
@@ -135,7 +135,7 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         performUpdatesNow()
     }
 
-    private fun findReactRootViewChildVM(fragmentManager: FragmentManager): FragmentManager? {
+    private fun findReactRootViewChildVM(fragmentManager: FragmentManager): FragmentManager {
         for (fragment in fragmentManager.fragments) {
             return if (fragment.view is ReactRootView) {
                 fragment.childFragmentManager
@@ -143,7 +143,7 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
                 findReactRootViewChildVM(fragment.childFragmentManager)
             }
         }
-        return null;
+        return fragmentManager;
     }
 
     private fun setupFragmentManager() {
@@ -184,7 +184,7 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         // whose root view is ReactRootView. As of now, we detect such case by checking whether any fragments are attached
         // to activity which hosts ReactRootView.
         if (context.supportFragmentManager.fragments.isNotEmpty()) {
-            setFragmentManager(findReactRootViewChildVM(context.supportFragmentManager)!!)
+            setFragmentManager(findReactRootViewChildVM(context.supportFragmentManager))
         } else {
             setFragmentManager(context.supportFragmentManager)
         }

--- a/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
+++ b/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
@@ -143,6 +143,8 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
                 findReactRootViewChildVM(fragment.childFragmentManager)
             }
         }
+
+        // In case there are no hosted fragments we just return passed fragment manager
         return fragmentManager;
     }
 
@@ -181,13 +183,9 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         check(context is FragmentActivity) { "In order to use RNScreens components your app's activity need to extend ReactActivity" }
 
         // In case React Native is loaded on a Fragment (not directly in activity) we need to find fragment manager
-        // whose root view is ReactRootView. As of now, we detect such case by checking whether any fragments are attached
+        // whose fragment's view is ReactRootView. As of now, we detect such case by checking whether any fragments are attached
         // to activity which hosts ReactRootView.
-        if (context.supportFragmentManager.fragments.isNotEmpty()) {
-            setFragmentManager(findReactRootViewChildVM(context.supportFragmentManager))
-        } else {
-            setFragmentManager(context.supportFragmentManager)
-        }
+        setFragmentManager(findReactRootViewChildVM(context.supportFragmentManager))
     }
 
     protected fun createTransaction(): FragmentTransaction {

--- a/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
+++ b/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
@@ -177,11 +177,11 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         // If parent is of type Screen it means we are inside a nested fragment structure.
         // Otherwise we expect to connect directly with root view and get root fragment manager
         if (parent is Screen) {
-            val screenFragment = parent.fragment
-            checkNotNull(screenFragment) { "Parent Screen does not have its Fragment attached" }
-            mParentScreenFragment = screenFragment
-            screenFragment.registerChildScreenContainer(this)
-            setFragmentManager(screenFragment.childFragmentManager)
+            checkNotNull(parent.fragment?.let { screenFragment ->
+                mParentScreenFragment = screenFragment
+                screenFragment.registerChildScreenContainer(this)
+                setFragmentManager(screenFragment.childFragmentManager)
+            }) { "Parent Screen does not have its Fragment attached" }
         } else {
             // we expect top level view to be of type ReactRootView, this isn't really necessary but in
             // order to find root view we test if parent is null. This could potentially happen also when

--- a/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
+++ b/android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt
@@ -141,7 +141,7 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
                 return if (fragment.view is ReactRootView) {
                     fragment.childFragmentManager
                 } else {
-                    findReactRootViewChildVM(fragment.childFragmentManager)
+                    _findReactRootViewChildVM(fragment.childFragmentManager)
                 }
             }
             return null

